### PR TITLE
net: socket: syscall for socketpair(2)

### LINKED ISSUE #24366: syscall for socketpair(2)
**Is your enhancement proposal related to a problem? Please describe.**

A typical means of inter-thread communication is to use a [socket pair](https://pubs.opengroup.org/onlinepubs/9699919799/functions/socketpair.html).

Although it may be a slightly heavyweight method of inter-thread communication for Zephyr, it would facilitate porting applications.  

**Describe the solution you'd like**
I would like to see an implementation of [socketpair(2)](https://pubs.opengroup.org/onlinepubs/9699919799/functions/socketpair.html) in Zephyr. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by cfriedt**
Provide an implementation for socketpair(2). Currently, this is limited to the AF_UNIX address family, SOCK_STREAM, SOCK_RAW, and protocol 0.

Please feel free to suggest additional testing in `tests/net/socket/socketpair/` or to suggest improvements in the implementation.

Fixes #24366

### CODE REVIEW

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+
+	int res;
+
+	/* Use a temporary socket vector since sv[] should not be modified
+         * on failure. See POSIX.1-2016 */
+	int temp_sv[2];
+
+	/* Zero-initializing this (with the exception of setting the family)
+	 * has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
+         *
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:188]: Some indentation issue on this line

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+	}
+
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
+		return -1;
+	}
+	temp_sv[0] = res;
+
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:222]: Please write this `if (res == -1) {` in order to follow the same style as the other functions in this file.

File: subsys/net/lib/sockets/sockets.c [hunk 3]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+		break;
+	default:
+		errno = EAFNOSUPPORT;
+		return -1;
+	}
+
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
+		return -1;
+	}
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:218]: Please add empty line after }

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+
+	res = z_impl_zsock_connect(temp_sv[0], sa, len);
+	if (-1 == res) {
+		z_impl_zsock_close(temp_sv[0]);
+		z_impl_zsock_close(temp_sv[1]);
+		return -1;
+	}
+
+	sv[ 0 ] = temp_sv[ 0 ];
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets.c:242]: no spaces inside `[]` so writing this as `sv[0] = temp_sv[0]` is ok

File: tests/net/socket/socketpair/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (jukkar) [inline tests/net/socket/socketpair/CMakeLists.txt:1]: Please put the tests to separate commit

File: tests/net/socket/socketpair/src/main.c [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_test, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <stdio.h>
+#include <string.h>
+#include <ztest_assert.h>
+
+#include <net/socket.h>
+
+//#include "../../socket_helpers.h"
```
Comment (jukkar) [inline tests/net/socket/socketpair/src/main.c:17]: Please remove this commented out line

File: tests/net/socket/socketpair/src/main.c [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+#ifndef ARRAY_SIZE
```
Comment (jukkar) [inline tests/net/socket/socketpair/src/main.c:19]: just include sys/util.h which will provide ARRAY_SIZE

File: tests/net/socket/socketpair/src/main.c [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	zassert_true(1, "1 was false");
+}
+
+void test_main(void)
+{
+	k_thread_system_pool_assign(k_current_get());
+
+	ztest_test_suite(socket_socketpair,
+			 ztest_user_unit_test(test_socket_socketpair_open_close));
+
```
Comment (jukkar) [inline tests/net/socket/socketpair/src/main.c:74]: We need to add a test that sends/receives data and makes sure that the data is passed properly.

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good, minor nits. We also need to add more tests.

File: subsys/net/lib/sockets/sockets.c [hunk 9]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+int z_impl_zsock_socketpair(int family, int type, int proto, int sv[2]) {
+
+	int res;
+
+	/* Use a temporary socket vector since sv[] should not be modified
+         * on failure. See POSIX.1-2016 */
+	int temp_sv[2];
+
+	/* Zero-initializing this (with the exception of setting the family)
+	 * has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c:187]: > Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.

Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this is required, or let me conjecture that this is serious security issue, where you open your "inter-thread communication" (per https://github.com/zephyrproject-rtos/zephyr/issues/24366) to attacks from the wide Internet.

Actually, you unlikely to provide such a reference, because normative spec says essebtially the opposite:

> The socketpair() function shall create an **unbound** pair of connected sockets

(emphasis mine). So, let me give -1 to this for now, and let's think how to deal with that.

File: subsys/net/lib/sockets/sockets.c [hunk 10]
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
+		z_impl_zsock_close(temp_sv[0]);
+		return -1;
+	}
+	temp_sv[1] = res;
+
+	res = z_impl_zsock_bind(temp_sv[1], sa, len);
+	if (-1 == res) {
+		z_impl_zsock_close(temp_sv[0]);
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c:230]: IMHO, repetitive close() call should be replaced with "C exception hanling with goto" pattern.

**Review by pfalcon — CHANGES_REQUESTED** [review]
There's an obvious security issue with this patch, see inline comment.

File: subsys/net/lib/sockets/sockets.c (hunk 9 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:187]: Hmm... Interesting. Normally when I look up this man page, I see that it says "unnamed" rather than "unbound". E.g. in Linux, it only works with AF_UNIX.

Also, I just realized I'm missing a listen(2) call, which would normally be necessary for e.g. AF_INET / SOCK_STREAM so it likely would not work in those cases and it would also expose a race condition / security issue, unless there was a way, as you say, to connect two sockets without binding.

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:187]: I would actually prefer to only support AF_UNIX, but since there is no such support in Zephyr yet, I thought I should work with what is available.

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c:187]: > I would actually prefer to only support AF_UNIX

That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:

1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.

Feel free to give this idea a thought.

File: subsys/net/lib/sockets/sockets.c (hunk 10 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:230]: I prefer that too. Some people are really against goto though.
 Will rework.

File: subsys/net/lib/sockets/sockets.c (hunk 9 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:187]: @pfalcon - if you would prefer as well, I could remove AF_INET and focus on adding AF_UNIX support instead.

Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c:187]: > I could remove AF_INET and focus on adding AF_UNIX support instead. Its use could be limited to socketpair() at least initially

+1 to this

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c]: > > I would actually prefer to only support AF_UNIX
> 
> That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

true...

> As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:
> 
> 1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
> 2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.
> 
> Feel free to give this idea a thought.

Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr. Are there additional / special markers that need to be inserted in order to support SOCK_DGRAM?

File: include/net/net_ip.h [hunk 11]
```diff
@@ -208,6 +213,12 @@ struct sockaddr_ll_ptr {
 	u8_t        *sll_addr;    /* Physical-layer address  */
 };
 
+/** Socket address for UNIX domain sockets, pipes, etc */
+struct sockaddr_un {
+	sa_family_t sun_family;    /* AF_UNIX  */
+	char        sun_path[108]; /* Pathname */
```
Comment (jukkar) [inline include/net/net_ip.h:219]: I think we need to make this configurable via Kconfig option. The value 108 is the same as in Linux but for Zephyr it is consuming way too much memory. For socketpair use, the path is not used according to manual `Likewise, the two sockets created by socketpair(2) are unnamed.  When the address of an unnamed socket is returned, its length is sizeof(sa_family_t), and sun_path should not be inspected.`
We could make the default value for path length so that sockaddr_un and sockaddr_in6 would be the same size. This way there is some space for path and user can increase the value if needed.

File: subsys/net/lib/sockets/sockets.c (hunk 9 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr.

@cfriedt, before discussing this further, I'd like to get your reply to https://github.com/zephyrproject-rtos/zephyr/issues/24366#issuecomment-613965163 . Because again, if you want to just signal an event to poll(), then eventfd() is a better choice.

(OTOH, if you really want to support SOCK_DGRAM (in addition to natural choice of SOCK_STREAM), then might be then reusing existing network socket infrastructure actually makes sense after all (or it would be indeed a bunch of code duplication between network socket subsystem and non-network socket impl).)

File: include/net/net_ip.h (hunk 11 above)
Comment (cfriedt) [reply include/net/net_ip.h:219]: @jukkar - yeah, I agree, it is a bit much, and technically speaking, it is unnecessary for socketpair(2) or pipe(2). Would it be better to remove this definition, since named sockets are not even part of the proposal? Originally, I was going to use the sun_path as either an actual path or a string of random bytes to assist in matching fds, but both of those things are probably unnecessary.

File: subsys/net/lib/sockets/sockets.c (hunk 9 above)
Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c]: @pfalcon - I would be fine just supporting SOCK_STREAM (SOCK_RAW is even probably unnecessary and I don't even know if there is a difference to SOCK_STREAM in this case, but some people prefer it for AF_UNIX).

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c]: > > Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
> 
> Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this 

Originally, I was just thinking INADDR_LOOPBACK, but then saw that `lo` requires an additional CONFIG option, but yea.. let's just forget about AF_INET.

File: subsys/net/lib/sockets/sockets.c [hunk 12]
```diff
@@ -173,6 +175,46 @@ int z_impl_zsock_socket(int family, int type, int proto)
...
+	errno = ENOSYS;
+	goto out;
+
+	res = 0;
+
+out:
+	return res;
+}
+
 #ifdef CONFIG_USERSPACE
```
Comment (ceolin) [inline subsys/net/lib/sockets/sockets.c:218]: I think it is missing the verification handler for this syscall. If that's correct your test is probably no calling this syscall from usermode

Comment (ceolin) [inline subsys/net/lib/sockets/sockets.c:218]: I think it is missing the verification handler for this syscall. If that's correct your test is probably no calling this syscall from usermode

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c:218]: Good catch! I think I have it now.

File: subsys/net/lib/sockets/sockets.c [hunk 13]
```diff
@@ -184,6 +226,17 @@ static inline int z_vrfy_zsock_socket(int family, int type, int proto)
 #include <syscalls/zsock_socket_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
+#ifdef CONFIG_USERSPACE
+static inline int z_vrfy_zsock_socketpair(int family, int type, int proto, int sv[2])
+{
+	/* implementation call to net_context_get() should do all necessary
+	 * checking
+	 */
```
Comment (ceolin) [inline subsys/net/lib/sockets/sockets.c:234]: You have to check if the calling thread has permission to `sv`.  So, here you have to create a temporary copy, call the implementation with this copy and after it return you copy the memory back using `z_user_to_copy`

Comment (ceolin) [reply subsys/net/lib/sockets/sockets.c:234]:
```
int ret;
int tmp[2];

ret = z_impl_zsock_socketpair(family, type, proto, tmp);
Z_OOPS(z_user_to_copy(sv, tmp, sizeof(tmp));
return ret;
```

Comment (cfriedt) [reply subsys/net/lib/sockets/sockets.c]: Thanks @ceolin !

**Comment by cfriedt** [comment]
If it's ok with you all (@pfalcon, @jukkar, @ceolin, @tbursztyka), I will leave this PR open and implement pipe(2) separately. I believe this review will be fairly cut & dry once pipe(2) is implemented.

Also, do you know of a simple way I can run my tests natively or in Qemu? So far all I have been contributing to is BLE.

Sorry for all of the whitespace issues - clang-format does not do the trick (#23823) but uncrustify might (#21392)

**Comment by cfriedt** [comment]
It's probably better to not use pipe(2) for socketpair(2) since a pair of sockets form a bidirectional channel rather than a unidirectional channel. Likely, will use a pair of k_fifo. Technically, it is possible using pipe(2) but that would mean 4 file descriptors would need to be created instead of only 2.

**Comment by pfalcon** [comment]
@cfriedt: Same comment as for the other PR - please deal with checkstyle issues.

(Also, to set expectations right - there was a talk about release readiness meeting. Which reminds that we may be approaching a pre-release freeze. Whatever isn't merged before it, may need to wait a month or so until next devel window is open. Surely feel free to work at you pace. What I wanted to say is that if you don't get a prompt review (once CI passes), don't hesitate to ping me. Thanks.)

**Comment by cfriedt** [comment]
Going to close the PR for the next while. Will reopen soon.
